### PR TITLE
CDAP-6589 Support Impersonation for logs

### PR DESCRIPTION
**Pull request by rohitsinha54**
Issue: https://issues.cask.co/browse/CDAP-6589
Build: http://builds.cask.co/browse/CDAP-DUT4476
- Supports impersonation for logs

Note: 
- This needs to be tested on a cluster.

### CODE REVIEW

**Comment by anwar6953** [comment]
Changes LGTM.
@gsps1 is just validating the changes against a cluster.

**Comment by rohitsinha54** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT4476-4
ITM: http://builds.cask.co/browse/CDAP-ITM-28

**Comment by gsps1** [comment]
Tested this PR, fixed two issues in ( commit 1dd3d5f3bf3d95cc6b1cb4054dd39d7c4e32303f)
1) skip any impersonation on system namespace in impersonator, log saver failed with impersonating access to log meta, but this is common to other places, so added it in Impersonator
2) AvroLogReader does read on location hdfs file, which has to be impersonated.

**Comment by rohitsinha54** [comment]
I had the AvroFileReader impersonation here in earlier commit:
https://github.com/caskdata/cdap/commit/9bfa05c9ac9ffc0787a190647f091ebd003b3379#diff-0af24fbc30d37cb485da80f8cf25884bR93
But removed it since we don't need impersonation during read/write and only when creating location. @anwar6953 : Why is this needed in this case ?

**Comment by rohitsinha54** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT4476-6
ITM: http://builds.cask.co/browse/CDAP-ITM-31

**Comment by anwar6953** [comment]
We need impersonation whenever creating the Location object, not just when creating the underlying file.
So even for reading a file, the location object needs to be created with impersonation.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileLogReader.java [hunk 1]
```diff
@@ -91,8 +90,7 @@ public void getLogNext(final LoggingContext loggingContext, final ReadRange read
 
       List<Location> filesInRange = getFilesInRange(sortedFiles, readRange.getFromMillis(), readRange.getToMillis());
       AvroFileReader logReader = new AvroFileReader(schema);
-      NamespaceId namespaceId =
-        new NamespaceId(loggingContext.getSystemTagsMap().get(NamespaceLoggingContext.TAG_NAMESPACE_ID).getValue());
+      NamespaceId namespaceId = LoggingContextHelper.getNamespaceId(loggingContext);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileLogReader.java:93]: nice, i didn't know about this class.

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java [hunk 2]
```diff
@@ -129,6 +130,7 @@ public static void initialize() throws IOException {
     metricsCollectionService = injector.getInstance(MetricsCollectionService.class);
 
     txManager = injector.getInstance(TransactionManager.class);
+    impersonator = injector.getInstance(Impersonator.class);
```
Comment (gsps1) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java:133]: where do you create the binding for impersonator in test module ?

**Comment by gsps1** [comment]
just one ( comment on 202f7c7d9478435b50fe0837c24c2c07e283d223) lgtm otherwise

**Comment by anwar6953** [comment]
LGTM

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java (hunk 2 above)
Comment (rohitsinha54) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java:133]: It is in the test module from earlier commit

**Pull request merged**